src/error.rs

```diff
@@ -7,13 +7,12 @@ pub enum Error {
     /// Application can authenticate against GCP using:
     ///
     /// - Default service account - available inside GCP platform using GCP Instance Metadata server
-    /// - Service account file - provided using `GOOGLE_APPLICATION_CREDENTIALS` with path
     /// - GCloud authorized user - retrieved using `gcloud auth` command
     ///
     /// All authentication methods have been tested and none succeeded.
     /// Service account file can be downloaded from GCP in json format.
     #[error("No available authentication method was discovered")]
-    NoAuthMethod(Box<Error>, Box<Error>, Box<Error>, Box<Error>),
+    NoAuthMethod(Box<Error>, Box<Error>, Box<Error>),
```

Comment:
It looks a potential error from `CustomServiceAccount` in `init()` no longer gets reported at all, now, which doesn't seem great. Maybe stick with the design I proposed in https://github.com/hrvolapeter/gcp_auth/issues/45#issuecomment-1005661685?

Reply:
I don't think that is the case. As I understand, this error is used to aggregate all errors in this function:

```rust
async fn get_authentication_manager(
    custom: Option<CustomServiceAccount>,
) -> Result<AuthenticationManager, Error> {
    #[cfg(feature = "webpki-roots")]
    let https = HttpsConnectorBuilder::new().with_webpki_roots();
    #[cfg(not(feature = "webpki-roots"))]
    let https = HttpsConnectorBuilder::new().with_native_roots();

    let client =
        Client::builder().build::<_, hyper::Body>(https.https_or_http().enable_http2().build());

    if let Some(service_account) = custom {
        log::debug!("Using CustomServiceAccount");
        return Ok(AuthenticationManager::new(
            client,
            Box::new(service_account),
        ));
    }
    let gcloud = gcloud_authorized_user::GCloudAuthorizedUser::new().await;
    if let Ok(service_account) = gcloud {
        log::debug!("Using GCloudAuthorizedUser");
        return Ok(AuthenticationManager::new(
            client.clone(),
            Box::new(service_account),
        ));
    }
    let default = default_service_account::DefaultServiceAccount::new(&client).await;
    if let Ok(service_account) = default {
        log::debug!("Using DefaultServiceAccount");
        return Ok(AuthenticationManager::new(
            client.clone(),
            Box::new(service_account),
        ));
    }
    let user = default_authorized_user::DefaultAuthorizedUser::new(&client).await;
    if let Ok(user_account) = user {
        log::debug!("Using DefaultAuthorizedUser");
        return Ok(AuthenticationManager::new(client, Box::new(user_account)));
    }
    Err(Error::NoAuthMethod(
        Box::new(gcloud.unwrap_err()),
        Box::new(default.unwrap_err()),
        Box::new(user.unwrap_err()),
    ))
}
```

I have removed the reading / parsing of the `CustomServiceAccount` outside this function, so that in my opinion the code becomes more flexible, as it no longer cares where the `CustomServiceAccount` comes from (as you can pass a `Option<CustomServiceAccount>`). Therefore this function no longer needs to report the error when the `CustomServiceAccount` fails to read the file / fails to parses the JSON.

* In case of init with a path to a file, the error is returned by `from_credentials_file`.
* In case of init using a json string, the error is returned by `from_credentials_json`.
* In case `GOOGLE_APPLICATION_CREDENTIALS` is set, but the file does not exist or can't be parsed, the error is returned by `init`.


This might actually solve one potential issue:

In case when a path is explicitly specified or when the `GOOGLE_APPLICATION_CREDENTIALS` is set and it fails to load or parse, I would expect it to fail immediately in case of an error, not to continue with detecting the other credentials options. This can be confusing if you are trying to override the credentials provided by the GCP environment.

Reply:
Oh, you're right -- sorry, I overlooked the `?` that you added after the `CustomServiceAccount::from_file(Path::new(&path)).await` call. I agree that this is an improvement!